compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs

```diff
@@ -0,0 +1,72 @@
+use rustc_middle::mir::interpret::Allocation;
+use rustc_middle::ty::layout::LayoutCx;
+use rustc_middle::ty::{ParamEnv, Ty, TyCtxt};
+use rustc_span::symbol::sym;
+use rustc_target::abi::{Align, Size};
+use std::mem::size_of;
+
+fn gc_layout(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, param_env: ParamEnv<'tcx>) -> Vec<u8> {
+    let layout = tcx.layout_of(param_env.and(ty)).unwrap();
+    let w_size = Size::from_bytes(size_of::<usize>());
+    let w_align = Align::from_bytes(w_size.bytes()).unwrap();
+
+    if layout.align.abi.bytes() < (w_size.bytes()) {
+        // If an ADT is aligned to a size smaller than 1 word, then it
+        // contains no fields large enough to represent a GC pointer.
+        //
+        // By default, the collector will conservatively scan a struct's
+        // memory, treating it as word-sized chunks. However, since rust is
+        // free to shuffle field order around, any in-memory words which
+        // represent actual pointers would be purely accidental. Such
+        // structs should be ignored completely by the collector.
+        return vec![0; 16]; // empty (bitmap, size) (u64,u64) as byte stream
+    }
+
+    let src_field_iter = ty.ty_adt_def().unwrap().all_fields();
+    let mut bitmap: u64 = 0xFFFFFFFF;
+
+    for i in layout.fields.index_by_increasing_offset() {
+        let target_offset = layout.fields.offset(i as usize);
```

Comment:
Is the `as` safe?

Reply:
Yes, the iterator will return nothing bigger than `u32`.

Reply:
I think we've gradually come to the conclusion that we want to avoid `as` if there's another mechanism. `usize::from(i)` if possible or `usize::try_from(i).unwrap()` otherwise (we assume the compiler removes the costs of the latter if it's infallible).

Reply:
While I agree in principle, rustc is littered with `as` casting of this kind. In fact this code was lifted from an identical pattern in the llvm codegen phase. I could buy this argument for our own crates, but here it feels like shutting the stable door after the horse has bolted.

Reply:
That's because most of the code in rustc predates `from`/`try_from`. I don't think there's a good reason for us to perpetuate that mistake, especially because `as` is so dangerous.

Reply:
Ok, fixed here 6b5b7565856b618372289638d57424354e71f288